lib/TH/generic/THTensorMath.c

```diff
@@ -72,6 +72,53 @@ void THTensor_(maskedSelect)(THTensor *tensor, THTensor *src, THByteTensor *mask
                    });
 }
 
+// Finds non-zero elements of a tensor and returns their subscripts
+void THTensor_(nonzero)(THLongTensor *subscript, THTensor *tensor)
+{
+  long numel = 0;
+  long *subscript_data;
+  long i = 0;
+  long dim;
+  long sub = 0;
+  long div = 1;
+
+  /* First Pass to determine size of subscripts */
+  TH_TENSOR_APPLY(real, tensor,
+                  if (*tensor_data != 0)
+                  {
+                    ++numel;
+                  });
+  THLongTensor_resize2d(subscript, numel, tensor->nDimension);
+
+  /* Second pass populates subscripts */
+  // The funky pointer arithmetic is as a result of the torch's last dimension
+  // being the fastest changing. Another valid strategy would be to fill in the
+  // subscript tensor backwards which would save some addition operations
+  // (2*num_non_zero_elems). However this method gives a result in numerical
+  // order which is more satisfying.
+  subscript_data = THLongTensor_data(subscript);
+  TH_TENSOR_APPLY(real, tensor,
+                  if (*tensor_data != 0)
+                  {
+                    subscript_data += tensor->nDimension - 1;
+
+                    *subscript_data = i % tensor->size[tensor->nDimension - 1];
+                    sub = *subscript_data;
+                    div = tensor->size[tensor->nDimension - 1];
+                    --subscript_data;
```

Comment:
I'd move these four lines into the loop to make the code more compact. If you use a temporary pointer to fill subscript_data you don't have to jump back and forth with subscript_data itself.